### PR TITLE
Add raw java example for `spi-based-sharding-algorithm-example`

### LINKED ISSUE #15717: Add raw java example for `spi-based-sharding-algorithm-example`
Hi community,

### Background

[spi-based-sharding-algorithm-example](https://github.com/apache/shardingsphere/tree/master/examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example) is an example of a custom algorithm that demonstrates how to integrate a user-defined sharding algorithm and configure it using  [SPI (Service Provider Interface)](https://docs.oracle.com/javase/tutorial/sound/SPI-intro.html).

### Aim

Add raw java example for `spi-based-sharding-algorithm-example` (in sub-module `spi-based-sharding-raw-jdbc-example`)

### Basic Qualifications

- Java
- Maven

### Example FYI

- [sharding-raw-jdbc-example](https://github.com/apache/shardingsphere/tree/master/examples/shardingsphere-jdbc-example/single-feature-example/sharding-example/sharding-raw-jdbc-example)

### PR DESCRIPTION
**Pull request by lushaorong**
Fixes #15717.

Changes proposed in this pull request:
- Add raw java example for `spi-based-sharding-algorithm-example`

### CODE REVIEW

File: examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/SPIBasedShardingRawJavaConfigurationExample.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+import javax.sql.DataSource;
+import java.sql.SQLException;
+
+public final class SPIBasedShardingRawJavaConfigurationExample {
+    
+    private static ShardingType shardingType = ShardingType.SHARDING_DATABASES;
+    
+    public static void main(final String[] args) throws SQLException {
+        DataSource dataSource = DataSourceFactory.newInstance(shardingType);
```
Comment (strongduanmu) [inline examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/SPIBasedShardingRawJavaConfigurationExample.java:34]: Do you think inline `ShardingType.SHARDING_DATABASES` is better?

File: examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/config/SPIBasedShardingRawJavaConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import org.apache.shardingsphere.sharding.api.config.rule.ShardingTableRuleConfiguration;
+import org.apache.shardingsphere.sharding.api.config.strategy.keygen.KeyGenerateStrategyConfiguration;
+import org.apache.shardingsphere.sharding.api.config.strategy.sharding.StandardShardingStrategyConfiguration;
+
+import javax.sql.DataSource;
+import java.sql.SQLException;
+import java.util.*;
+
+public final class SPIBasedShardingRawJavaConfiguration  implements ExampleConfiguration {
+    @Override
```
Comment (strongduanmu) [inline examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/config/SPIBasedShardingRawJavaConfiguration.java:34]: Please add a blank line before here.

File: examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/config/SPIBasedShardingRawJavaConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        result.setKeyGenerateStrategy(new KeyGenerateStrategyConfiguration("order_id", "snowflake"));
+        return result;
+    }
+
+    private ShardingTableRuleConfiguration getOrderItemTableRuleConfiguration() {
+        ShardingTableRuleConfiguration result = new ShardingTableRuleConfiguration("t_order_item", "demo_ds_${0..1}.t_order_item_${[0, 1]}");
+        result.setKeyGenerateStrategy(new KeyGenerateStrategyConfiguration("order_item_id", "snowflake"));
+        return result;
+    }
+
```
Comment (strongduanmu) [inline examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/config/SPIBasedShardingRawJavaConfiguration.java:70]: Please remove this blank line.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/SPIBasedShardingRawJavaConfigurationExample.java (hunk 1 above)
Comment (lushaorong) [reply examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/SPIBasedShardingRawJavaConfigurationExample.java:34]: I refer to [SPIBasedShardingRawYamlConfigurationExample](https://github.com/apache/shardingsphere/blob/master/examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/SPIBasedShardingRawYamlConfigurationExample.java)

File: examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/config/SPIBasedShardingRawJavaConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import org.apache.shardingsphere.example.core.api.DataSourceUtil;
+import org.apache.shardingsphere.infra.config.algorithm.ShardingSphereAlgorithmConfiguration;
+import org.apache.shardingsphere.sharding.api.config.ShardingRuleConfiguration;
+import org.apache.shardingsphere.sharding.api.config.rule.ShardingTableRuleConfiguration;
+import org.apache.shardingsphere.sharding.api.config.strategy.keygen.KeyGenerateStrategyConfiguration;
+import org.apache.shardingsphere.sharding.api.config.strategy.sharding.StandardShardingStrategyConfiguration;
+
+import javax.sql.DataSource;
+import java.sql.SQLException;
+import java.util.*;
```
Comment (RaigorJiang) [inline examples/shardingsphere-jdbc-example/single-feature-example/extension-example/custom-sharding-algortihm-example/spi-based-sharding-algorithm-example/spi-based-sharding-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/extension/spibased/sharding/raw/jdbc/config/SPIBasedShardingRawJavaConfiguration.java:31]: Please do not use `*` to import dependencies.
If using IDEA, you can import the recommended [Settings](https://shardingsphere.apache.org/community/data/shardingsphere-settings.jar).

Refer to: https://shardingsphere.apache.org/community/en/contribute/code-conduct/

**Comment by RaigorJiang** [comment]
Hi @lushaorong 
CI failed because the master branch has been refactored, you need to merge the latest master code in your own branch and push again, thank you!

**Pull request closed without merging**